optimum/habana/transformers/models/gemma2/modeling_gemma2.py

```diff
@@ -0,0 +1,1005 @@
+        use_new_cache = False  # Ignoring new Cache path for HPU
+
+        past_seen_tokens = 0
+
+        if past_key_values is not None and use_cache:  # kept for BC (cache positions)
+            if reuse_cache:
+                if isinstance(past_key_values[0][0], torch.Tensor):
+                    past_seen_tokens = past_key_values[0][0].shape[2]
+                else:
+                    past_seen_tokens = past_key_values[0][0][2]
+            else:
+                if use_new_cache:
+                    if not isinstance(past_key_values, StaticCache):
+                        past_key_values = DynamicCache.from_legacy_cache(past_key_values)
+                    past_seen_tokens = past_key_values.get_seq_length()
+                else:
+                    past_seen_tokens = past_key_values[0][0].shape[2]
+
+        if ignore_cache_position is False:
+            if cache_position is None:
+                past_seen_tokens = past_key_values.get_seq_length() if past_key_values is not None else 0
+                cache_position = torch.arange(
+                    past_seen_tokens, past_seen_tokens + inputs_embeds.shape[1], device=inputs_embeds.device
+                )
+            if position_ids is None and cache_position:
+                position_ids = cache_position.unsqueeze(0)
+        else:
+            if position_ids is None:
+                position_ids = torch.arange(
+                    past_seen_tokens, seq_length + past_seen_tokens, dtype=torch.long, device=inputs_embeds.device
+                )
+                position_ids = position_ids.unsqueeze(0)
+            cache_position = None
+
+        # HPU specific mask generation
+        if ignore_cache_position:
+            causal_mask = _gaudi_prepare_4d_causal_attention_mask(
+                attention_mask,
+                input_ids.shape if input_ids is not None else (batch_size, seq_length),
+                inputs_embeds,
+                past_seen_tokens,
+            )
+        else:
+            causal_mask = self._update_causal_mask(attention_mask, inputs_embeds, cache_position, past_seen_tokens)
+
+        # embed positions
+        hidden_states = inputs_embeds
+
+        normalizer = torch.tensor(self.config.hidden_size**0.5, dtype=hidden_states.dtype, device=inputs_embeds.device)
+        hidden_states = hidden_states * normalizer
+
+        # decoder layers
+        all_hidden_states = () if output_hidden_states else None
+        all_self_attns = () if output_attentions else None
+        next_decoder_cache = () if not use_new_cache else None
+
+        if lazy_mode:
+            htcore.mark_step()
+
+        for layer_idx, decoder_layer in enumerate(self.layers):
+            if (
+                lazy_mode
+                and not self.training
+                and (torch.distributed.is_initialized() is False or torch.distributed.get_world_size() == 1)
+            ):
+                htcore.mark_step()
```

Comment:
Some models have this every layer: https://github.com/huggingface/optimum-habana/blob/main/optimum/habana/transformers/models/llama/modeling_llama.py#L1137


Some have it every 2 layer:
https://github.com/huggingface/optimum-habana/blob/main/optimum/habana/transformers/models/mistral/modeling_mistral.py


Some do not have it:
https://github.com/huggingface/optimum-habana/blob/main/optimum/habana/transformers/models/falcon/modeling_falcon.py#L889


Can you please check which config gives you best performance, if you haven't already

Reply:
Checked and llama-way (used in my patch) seems to be trade-off way to achieve best performace meanwhile reduce memory occupation.

```
python run_generation.py --model_name_or_path ../../../gemma-2-9b --bf16 --trim_logits --use_kv_cache --use_hpu_graphs --use_flash_attention --reuse_cache --max_input_tokens 128 --max_new_tokens 128 --batch_size 256

Stats:---------------------------------------------------------------------------------------------------------------
Throughput (including tokenization) = 5976.0827954356755 tokens/second
Number of HPU graphs                = 20
Memory allocated                    = 44.48 GB
Max memory allocated                = 50.16 GB
Total memory available              = 94.62 GB
Graph compilation duration          = 18.32479699200485 seconds
---------------------------------------------------------------------------------------------------------------

(remove markstep)
root@800672-3:/intel/optimum-habana-gemma/examples/text-generation# python run_generation.py --model_name_or_path ../../../gemma-2-9b --bf16 --trim_logits --use_kv_cache --use_hpu_graphs --use_flash_attention --reuse_cache --max_input_tokens 128 --max_new_tokens 128 --batch_size 256

Stats:--------------------------------------------------------------------------------------------------------------
Throughput (including tokenization) = 6011.527381214152 tokens/second
Number of HPU graphs                = 18
Memory allocated                    = 44.26 GB
Max memory allocated                = 94.44 GB (huge memory allocated!!!)
Total memory available              = 94.62 GB
Graph compilation duration          = 61.71153533604229 seconds
--------------------------------------------------------------------------------------------------------------

```